GVFS/GVFS.Common/Tracing/QueuedPipeStringWriter.cs

```diff
@@ -0,0 +1,208 @@
+        private int state = (int)QueuedPipeStringWriterState.Unknown;
+        private bool isDisposed;
+
+        public QueuedPipeStringWriter(Func<NamedPipeClientStream> createPipeFunc, int maxQueueSize = DEFAULT_MAX_QUEUE_SIZE)
+        {
+            this.createPipeFunc = createPipeFunc;
+            this.queue = new BlockingCollection<string>(new ConcurrentQueue<string>(), boundedCapacity: maxQueueSize);
+        }
+
+        public event EventHandler<QueuedPipeStringWriterStateEventArgs> StateChanged;
+
+        public enum QueuedPipeStringWriterState
+        {
+            Unknown = 0,
+            Stopped = 1,
+            Failing = 2,
+            Healthy = 3,
+        }
+
+        public void Start()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (this.writerThread != null)
+            {
+                return;
+            }
+
+            this.writerThread = new Thread(this.ThreadProc)
+            {
+                Name = nameof(QueuedPipeStringWriter),
+                IsBackground = true,
+            };
+
+            this.writerThread.Start();
+        }
+
+        public bool TryEnqueue(string message)
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            return this.queue.TryAdd(message);
+        }
+
+        public void Stop()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (this.queue.IsAddingCompleted)
+            {
+                return;
+            }
+
+            // Signal to the queue draining thread that it should drain once more and then terminate.
+            this.queue.CompleteAdding();
+            this.writerThread.Join();
+
+            Debug.Assert(this.queue.IsCompleted, "Message queue should be empty after being stopped");
+        }
+
+        public void Dispose()
+        {
+            if (!this.isDisposed)
+            {
+                this.Stop();
+
+                this.pipeClient?.Dispose();
+                this.pipeClient = null;
+
+                this.writerThread = null;
+            }
+
+            this.isDisposed = true;
+        }
+
+        private void RaiseStateChanged(QueuedPipeStringWriterState newState, Exception ex)
+        {
+            int prevState = Interlocked.Exchange(ref this.state, (int)newState);
+
+            if (prevState != (int)newState)
+            {
+                this.StateChanged?.Invoke(this, new QueuedPipeStringWriterStateEventArgs(newState,  ex));
+            }
+        }
+
+        private void ThreadProc()
+        {
+            // Drain the queue of all messages currently in the queue
```

Comment:
> [..] with a call to Environment.Exit in the catch.

I don't think we want to bring the mount process down for this sort of failure in telemetry?

Reply:
>I don't think we want to bring the mount process down for this sort of failure in telemetry?

With the current code what would happen to the mount process if there were an unhandled exception in `ThreadProc`?

Reply:
I mean, yes it would bring down the process as is. I agree with want the `try/catch`.
What I'm not sure about is the `Environment.Exit(1)`.. I don't think we should terminate the entire mount process because the telemetry background thread hit a problem.

This raises another question about recovery there then too.. should this thread attempt recovery?

Reply:
\+ @jrbriggs for his thoughts on this as well.

In general how much should VFS4G explode if it expects telemetry to be working and finds that it's not?